### PR TITLE
Testing

### PR DESCRIPTION
**Pull request by michamichaa**
Pull Request for ready to test Teleop code. All subsystems already merged here

### CODE REVIEW

File: 2024Robot_Team3484/src/main/cpp/subsystems/IntakeSubsystem.cpp [hunk 1]
```diff
@@ -39,17 +39,21 @@ IntakeSubsystem::IntakeSubsystem(
...
     _pivot_pid_controller->SetFF(PID_CONSTANTS.Kf);
-    _pivot_pid_controller->SetOutputRange(PID_OUTPUTRANGE_MIN, PID_OUTPUTRANGE_MAX);
+    _pivot_pid_controller->SetOutputRange(pid_output_range_min, pid_output_range_max);
 }
 
 void IntakeSubsystem::Periodic() {
     // Runs every 20ms
 
+    #ifdef EN_DIAGNOSTICS
+        SmartDashboard::GetNumber("Intake Angle (deg)", _pivot_encoder->GetPosition()*360);
```
Comment (Alex-BusNet) [inline 2024Robot_Team3484/src/main/cpp/subsystems/IntakeSubsystem.cpp:54]: This should be `SmartDashboard::PutNumber( )` not `SmartDashboard::GetNumber( )`. Also I suggest adding a diagnostic for the pivot motor's velocity.

File: 2024Robot_Team3484/src/main/cpp/subsystems/LauncherSubsystem.cpp [hunk 2]
```diff
@@ -40,36 +40,56 @@ LauncherSubsystem::LauncherSubsystem(
...
+        return std::abs(Launcher_Encoder_Left->GetVelocity()-_target_speed) < RPM_WINDOW_RANGE && std::abs(Launcher_Encoder_Right->GetVelocity()-_target_speed) < RPM_WINDOW_RANGE;   
+    }
+    else if (_counter_null_left == 1) {
+        return std::abs(Launcher_Encoder_Left->GetVelocity()-_target_speed) < RPM_WINDOW_RANGE;   
+    }
+    else if (_counter_null_right == 1) {
+        return std::abs(Launcher_Encoder_Right->GetVelocity()-_target_speed) < RPM_WINDOW_RANGE;   
+    }
+    if (_counter_null_left + _counter_null_right == 2){
+        return false;
```
Comment (Alex-BusNet) [inline 2024Robot_Team3484/src/main/cpp/subsystems/LauncherSubsystem.cpp:93]: Based on how your null counter logic works in the periodic function. You will be returning `FALSE` when both PID controllers are NOT null, and the `atTargetRPM` function will only return `TRUE` when at least 1 of the PID controllers is null. To fix this you can either:
a) Change the first `if` condition to check if your null counters are equal to 2.
b) Move the counter increment statements in `Periodic()` to an `else` condition of the PID controller null checks.

```
if (both are not null)
{
    return true if both PID controllers are in range
}
else if (right is null)
{
    return true if left PID controller is in range
}
else if (left is null)
{
    return true if right PID controller is in range
}
else
{
    return false;
}
```

Also, I suggest having the `return false;` statement be an `else` condition in the `atTargetRPM` function.

File: 2024Robot_Team3484/src/main/include/subsystems/ClawSubsystem.h [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+            int right_sensor_di_ch
+        );
+        void Periodic() override;
+        void SetClawPower(double power);
+        bool GetLeftSensor();
+        bool GetRightSensor();
+
+    private: 
+        WPI_TalonSRX _left_climb_motor;
+        WPI_TalonSRX _right_climb_motor;
```
Comment (Alex-BusNet) [inline 2024Robot_Team3484/src/main/include/subsystems/ClawSubsystem.h:25]: These have been changed to SparkMax controllers.

File: 2024Robot_Team3484/src/main/cpp/commands/auton/AutonAimCommand.cpp
[the remark this answers is not in the document]
Comment (michamichaa) [reply 2024Robot_Team3484/src/main/cpp/commands/auton/AutonAimCommand.cpp:1]: yeah I believe the main difference is in the logic ti receive a true value in the whole if-else statement. In the Teleop Aim Command, we have the ability to override the aim sequence by pressing a button on the controller

File: 2024Robot_Team3484/src/main/cpp/commands/teleop/TeleopIntakeCommand.cpp [hunk 4]
```diff
@@ -1,18 +1,20 @@
...
         AddRequirements(_intake_subsystem);
+        AddRequirements(_launcher_subsystem);
 }
 
 void TeleopIntakeCommand::Initialize() {}
 
 void TeleopIntakeCommand::Execute() {
     if (_oi->ExtendIntakeButton()) {
-        if (!_intake_subsystem->HasPiece() || _oi->IntakeOverrideButton()) {
+        if ((!_intake_subsystem->HasPiece() || _oi->IgnoreVision()) && _oi != NULL) {
```
[the remark this answers is not in the document]
Comment (nfoert) [reply 2024Robot_Team3484/src/main/cpp/commands/teleop/TeleopIntakeCommand.cpp:17]: I moved the null check, and replaced `IgnoreVision` with the already defined `IgnoreSensor`. We were intending to use this to allow the intake to extend, no matter what value `HasPiece` is returning.

**Review by Alex-BusNet — APPROVED** [review]

**Pull request merged**